### PR TITLE
move on-refresh config to unit databag

### LINKED ISSUE #140: Grafana Agent Subordinate Errors out for non-leader units
### Bug Description

I'm deploying Kafka (on VMs) with multiple units (n=3). When I deploy and relate the grafana agent charm, the grafana-agent unit related to the Kafka juju leader unit works fine but the other two (related to the non-leaders units) errors out because of a permission error in the relation data bag. 

This is the output of the juju status

```
Model    Controller  Cloud/Region         Version  SLA          Timestamp
default  dev         localhost/localhost  2.9.42   unsupported  23:14:08Z

App            Version  Status  Scale  Charm          Channel  Rev  Exposed  Message
grafana-agent           active      3  grafana-agent  edge       2  no       
kafka                   active      3  kafka          edge     106  no       
zookeeper               active      3  zookeeper      edge      92  no       

Unit                Workload  Agent  Machine  Public address  Ports  Message
kafka/6*            active    idle   24       10.21.168.7            
  grafana-agent/2   active    idle            10.21.168.7            
kafka/7             error     idle   25       10.21.168.75           hook failed: "cos-agent-relation-joined"
  grafana-agent/0*  active    idle            10.21.168.75           
kafka/8             error     idle   26       10.21.168.72           hook failed: "cos-agent-relation-joined"
  grafana-agent/1   active    idle            10.21.168.72           
zookeeper/6         active    idle   27       10.21.168.30           
zookeeper/7*        active    idle   28       10.21.168.101          
zookeeper/8         active    idle   29       10.21.168.85           
```

### To Reproduce

1. juju deploy kafka -n3 --channel edge
2. juju deploy zookeeper -n3 --channel edge
3. juju relate kafka zookeeper
4. juju deploy grafana-agent --channel edge
5. (wait everything to be active)
6. juju relate grafana-agent kafka

### Environment

juju version 
2.9.42-ubuntu-amd64

lxd version 
5.11

### Relevant log output

```shell
unit-kafka-8: 23:13:51 ERROR unit.kafka/8.juju-log cos-agent:27: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-kafka-8/charm/./src/charm.py", line 503, in <module>
    main(KafkaCharm)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/main.py", line 436, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/main.py", line 144, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/framework.py", line 354, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/framework.py", line 830, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/framework.py", line 919, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-kafka-8/charm/lib/charms/grafana_agent/v0/cos_agent.py", line 251, in _on_refresh
    relation.data[self._charm.app].update({"config": self._generate_databag_content()})
  File "/usr/lib/python3.10/_collections_abc.py", line 994, in update
    self[key] = other[key]
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/model.py", line 1473, in __setitem__
    self._validate_write(key, value)
  File "/var/lib/juju/agents/unit-kafka-8/charm/venv/ops/model.py", line 1459, in _validate_write
    raise RelationDataAccessError(
ops.model.RelationDataAccessError: kafka/8 is not leader and cannot write application data.
unit-kafka-8: 23:13:51 ERROR juju.worker.uniter.operation hook "cos-agent-relation-joined" (via hook dispatching script: dispatch) failed: exit status 1
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lucabello**
## Issue
Closes #140.

## Solution
Move the config information to the unit databag instead of the application one.

## Testing Instructions
Pack this charm, add this library to the Kafka charm and pack it; scale up Kafka and then set up the `cos-agent` relation, then relating the agent to Prometheus. Verify that metrics for all units are in Prometheus.

### CODE REVIEW

**Comment by sed-i** [comment]
We currently have an [itest for `juju_info`](https://github.com/canonical/grafana-agent-k8s-operator/blob/machine_charm/tests/integration-machine/test_juju_info.py) with `num_units=2`. Could you please check if we need a leader guard for `juju_info`? If yes, then the test should be fixed because it didn't catch it.

Ideally we would have an itest for `cos-agent` as well, but kafka and zookeeper are too heavy.

**Comment by lucabello** [comment]
I think the leader guard discussion is now outdated since we decided to go with the unit-databag-approach instead!

Here you can see that my scaled-up Kafka using the new library correctly sends metrics to Prometheus for both units:
Screenshot_20230322_114319

File: lib/charms/grafana_agent/v0/cos_agent.py [hunk 1]
```diff
@@ -248,21 +248,21 @@ def _on_refresh(self, event):
...
+                if self._charm.unit.is_leader():
+                    relation.data[self._charm.app].update({"config": self._generate_application_databag_content()})
+                relation.data[self._charm.unit].update({"config": self._generate_unit_databag_content()})
 
-    def _generate_databag_content(self) -> str:
-        """Collate the data for each nested databag and return it."""
-        # The databag is divided in three chunks: metrics, logs, and dashboards.
+    def _generate_application_databag_content(self) -> str:
+        """Collate the data for each nested app databag and return it."""
+        # The application databag is divided in two chunks: metrics (alert rules only) and dashboards.
```
Comment (sed-i) [inline lib/charms/grafana_agent/v0/cos_agent.py:257]: **suggestion:** The comment doesn't match the code. Should it be:
```suggestion
        # The application databag is divided in three chunks: alert rules (metrics and logs) and dashboards.
        # Scrape jobs and log slots are unit-dependent and are therefore stored in unit databag.
```

File: lib/charms/grafana_agent/v0/cos_agent.py [hunk 2]
```diff
@@ -272,6 +272,23 @@ def _generate_databag_content(self) -> str:
 
         return json.dumps(data)
 
+    def _generate_unit_databag_content(self) -> str:
+        """Collate the data for each nested unit databag and return it."""
+        # The unit databag is divided in two chunks: metrics (scrape jobs only) and logs.
```
Comment (sed-i) [inline lib/charms/grafana_agent/v0/cos_agent.py:277]: **nitpick (non-blocking):**
```suggestion
        # The unit databag is divided in two chunks: metrics (scrape jobs only) and logs.
        # Alert rules and dashboards are unit-independent and are therefore stored in app databag.
```

File: lib/charms/grafana_agent/v0/cos_agent.py [hunk 3]
```diff
@@ -361,6 +378,13 @@ def trigger_refresh(self, _):
         # FIXME: Figure out what we should do here
         self.on.data_changed.emit()
 
+    @staticmethod
+    def _relation_unit(relation: Relation) -> Optional[Unit]:
+        """Return the principal unit for a relation."""
+        if relation and relation.units:
+            return next(iter(relation.units))
```
Comment (sed-i) [inline lib/charms/grafana_agent/v0/cos_agent.py:385]: **question:** Is the implied order guaranteed? I.e. is the principal guaranteed to be the first item in the list?

**Review by Abuelodelanada — APPROVED** [review]

**Comment by sed-i** [comment]
Post-merge comments implemented in #141

**Pull request merged**